Util/Xamarin.Build.Download/source/Xamarin.Build.Download/XamarinBuildAndroidAarProguardConfigs.cs

```diff
@@ -0,0 +1,123 @@
+			proguardIntermediateOutputPath = Path.Combine (MergeOutputDir, "proguard");
+			if (!Directory.Exists (proguardIntermediateOutputPath))
+				Directory.CreateDirectory (proguardIntermediateOutputPath);
+
+			var additionalFileWrites = new List<ITaskItem> ();
+
+			// Make sure our XbdMerge directory exists
+			var outputDir = MergeOutputDir;
+			Directory.CreateDirectory (outputDir);
+
+			// Get our assembly restore map
+			var restoreMap = BuildRestoreMap (RestoreAssemblyResources);
+			if (restoreMap == null)
+				return false;
+
+			IAssemblyResolver resolver = CreateAssemblyResolver ();
+
+			// Look through all the assemblies we would restore for
+			foreach (var asm in restoreMap) {
+
+				// We only want to find proguard files in .aar files referenced
+				// for assemblies we actually have referenced and have them mapped to
+				var asmName = new AssemblyName (asm.Key);
+				ITaskItem item = FindMatchingAssembly (InputReferencePaths, asmName);
+				if (item == null) {
+					if (ThrowOnMissingAssembly)
+						return false;
+					else
+						continue;
+				}
+
+				// Use a hash for the assembly name to keep paths shorter
+				var saveNameHash = DownloadUtils.HashMd5 (asmName.Name)?.Substring (0, 8);
+
+				// We keep a stamp file around to avoid reprocessing, so skip if it exists
+				var stampPath = Path.Combine (outputDir, saveNameHash + ".proguard.stamp");
+				if (File.Exists (stampPath))
+					continue;
+
+				// Get all the mapped .aar files
+				var resourceItems = asm.Value;
+
+				// We want to increment on the hash name in case there are multiple .aar files and/or proguard config
+				// files for a given assembly, so we use them all and not overwrite the same name
+				var entryCount = 0;
+
+				// In theory we could have multiple .aar files? Probably never happen...
+				foreach (var resourceItem in resourceItems) {
+					
+					// Full path to .aar file
+					var resourceFullPath = resourceItem.GetMetadata ("FullPath");
+
+					using (var fileStream = File.OpenRead (resourceFullPath))
+					using (var zipArchive = new ZipArchive (fileStream, ZipArchiveMode.Read)) {
+
+						// Look for proguard config files in the archive
+						foreach (var entry in zipArchive.Entries) {
+							
+							// Skip entries which are not proguard configs
+							if (!entry.Name.Equals ("proguard.txt", StringComparison.OrdinalIgnoreCase)
+								&& !entry.Name.Equals ("proguard.cfg", StringComparison.OrdinalIgnoreCase))
+								continue;
+
+							// Figure out our destination filename
+							var proguardSaveFilename = Path.Combine (proguardIntermediateOutputPath, saveNameHash + entryCount + ".txt");
+
+							// Save out the proguard file
+							using (var entryStream = entry.Open ())
+							using (var fs = File.Create (proguardSaveFilename)) {
+								entryStream.CopyTo (fs);
+								fs.Flush ();
+								fs.Close ();
+							}
+
+							entryCount++;
+						}
+					}
+
+					entryCount++;
```

Comment:
I may be nitpicking, but de we need to increment here? we are basically going through a collection of collections, and if we are incrementing each time in the inner loop, this technically will skip a number. I am just thinking that if we don't have to increment, then don't - we may be debugging this later and we may think that something is wrong when our files go 1, 2, 4, 5, 7, 8 (or something like that).